pkg/sql/schemachanger/sctest/comparator.go

```
@@ -0,0 +1,283 @@
+	}
+
+	return strings.Join(newLine, "; ")
+}
+
+// requireNoErrOrSameErrCode require errors from executing some statement
+// from legacy and declarative schema changer clusters to be both nil or
+// both PQ error with same code.
+func requireNoErrOrSameErrCode(t *testing.T, line string, errLegacy, errDeclarative error) {
+	if errLegacy == nil && errDeclarative == nil {
+		return
+	}
+
+	if errLegacy == nil {
+		t.Fatalf("statement %q failed with declarative schema changer (but succeeded with legacy schema changer): %v", line, errDeclarative.Error())
+	}
+	if errDeclarative == nil {
+		t.Fatalf("statement %q failed with legacy schema changer (but succeeded with declarative schema changer): %v", line, errLegacy.Error())
+	}
+	errLegacyPQCode := getPQErrCode(errLegacy)
+	errDeclarativePQCode := getPQErrCode(errDeclarative)
+	if errLegacyPQCode == "" || errDeclarativePQCode == "" {
+		t.Fatalf("executing statement %q results in non-PQ error:  legacy=%v, declarative=%v ", line, errLegacy.Error(), errDeclarative.Error())
+	}
+	if errLegacyPQCode != errDeclarativePQCode {
+		t.Fatalf("executing statement %q results in different error code: legacy=%v, declarative=%v", line, errLegacyPQCode, errDeclarativePQCode)
+	}
+}
+
+func getPQErrCode(err error) pq.ErrorCode {
+	if pqErr := (*pq.Error)(nil); errors.As(err, &pqErr) {
+		return pqErr.Code
+	}
+	return ""
+}
+
+// containsStmtOfType returns true if `line` contains any statement of type `typ`.
+func containsStmtOfType(t *testing.T, line string, typ tree.StatementType) bool {
+	parsedLine, err := parser.Parse(line)
+	require.NoError(t, err)
+	for _, parsedStmt := range parsedLine {
+		if parsedStmt.AST.StatementType() == typ {
+			return true
+		}
+	}
+	return false
+}
+
+// metaDataIdentityCheck looks up all descriptors' create_statements in
+// `legacy` and `declarative` clusters and assert that they are identical.
+func metaDataIdentityCheck(t *testing.T, legacy, declarative *sqlutils.SQLRunner) error {
```

nit: Having a function accept a `testing.T` and return an error is a bit difficult for readers to reason about. I think it would be more ergonomic to either not accept a testing.T and return errors for other operations or call t.Fatal/t.Fail rather than returning an error.